File: flutter_theolive_sdk_android/android/src/main/kotlin/com/theolive/flutter/THEOliveView.kt
package com.theolive.flutter

import android.content.Context
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import com.theolive.flutter.pigeon.PigeonNativePlayerConfiguration
import com.theolive.flutter.pigeon.THEOliveFlutterAPI
import com.theolive.flutter.pigeon.THEOliveNativeAPI
import com.theolive.player.EventListener
import com.theolive.player.PlayerView
import com.theolive.player.RenderingTarget
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.platform.PlatformView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class THEOliveView(context: Context, viewId: Int, args: Any?, messenger: BinaryMessenger) : PlatformView, EventListener, THEOliveNativeAPI {

    private val RENDERINGTARGET_SURFACE_VIEW = "surfaceView"

    private val id: Int
    private val pigeonMessenger: PigeonMultiInstanceBinaryMessengerWrapper
    private val flutterApi: THEOliveFlutterAPI
    private val linearLayout: LinearLayout
    private val playerView: PlayerView
    private val nativeRenderingTarget: String
    private val emptyCallback: (Result<Unit>) -> Unit = {}
    private val mainScope = CoroutineScope(Dispatchers.Main)

    // Workaround to eliminate the initial transparent layout with initExpensiveAndroidView
    // TODO: remove it once initExpensiveAndroidView is not used.
    private var useHybridComposition: Boolean = false
    private var isFirstPlaying: Boolean = false
        set(value) {
            if (!useHybridComposition) {
                return
            }

            if (value) {
                playerView.visibility = View.VISIBLE
            } else {
                playerView.visibility = View.INVISIBLE
            }
            field = value
        }

    // sync behavior with native iOS SDK
    // TODO: this logic should be fixed in the native SDK.
    private var wasPlayingBeforeActivityOnPause: Boolean = false

    init {
        id = viewId
        useHybridComposition = (args as? Map<*, *>)?.get("useHybridComposition") as? Boolean == true
        nativeRenderingTarget = ((args as? Map<*, *>)?.get("nativeRenderingTarget") as? String) ?: RENDERINGTARGET_SURFACE_VIEW //TODO: use enum and pigeon

        pigeonMessenger = PigeonMultiInstanceBinaryMessengerWrapper(messenger, "id_$viewId")
        THEOliveNativeAPI.setUp(pigeonMessenger, this)
        flutterApi = THEOliveFlutterAPI(pigeonMessenger)

        val layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        linearLayout = LinearLayout(context)
        linearLayout.layoutParams = layoutParams
        linearLayout.id = viewId
        playerView = PlayerView(context)
        playerView.id = View.generateViewId()
        playerView.layoutParams = layoutParams
        linearLayout.addView(playerView)

        if (nativeRenderingTarget == RENDERINGTARGET_SURFACE_VIEW) {
            playerView.setRenderingTarget(RenderingTarget.SURFACE_VIEW)
        } else {
            playerView.setRenderingTarget(RenderingTarget.TEXTURE_VIEW)
        }

        playerView.player.addEventListener(this)
    }

    override fun getView(): View {
        return linearLayout
    }

    override fun dispose() {
        // do nothing, we rely on destroy() method to clean things
    }

    override fun preloadChannels(channelIDs: List<String>) {
        playerView.player.preloadChannels(channelIDs.toTypedArray())
    }

    override fun loadChannel(channelID: String) {
        playerView.player.loadChannel(channelID)
    }

    override fun play() {
        playerView.player.play()
    }

    override fun pause() {
        playerView.player.pause()
    }

    override fun isAutoplay(): Boolean {
        return playerView.player.isAutoplay
    }

    override fun setMuted(muted: Boolean) {
        playerView.player.muted = muted
    }

    override fun setBadNetworkMode(badNetworkMode: Boolean) {
        playerView.player.badNetworkMode = badNetworkMode
    }

    override fun goLive() {
        playerView.player.goLive()
    }

    override fun reset() {
        playerView.player.reset()
    }

    override fun updateConfiguration(configuration: PigeonNativePlayerConfiguration) {
        playerView.player.updateConfig(configuration.sessionId)
    }

    override fun destroy() {
        playerView.player.removeEventListener(this)
        linearLayout.removeView(playerView)
        playerView.player.destroy()
        playerView.onDestroy()
    }

    override fun onLifecycleResume() {
        playerView.onResume()
        if (wasPlayingBeforeActivityOnPause) {
            wasPlayingBeforeActivityOnPause = false
            play()
        }
    }

    override fun onLifecyclePause() {
        wasPlayingBeforeActivityOnPause = !playerView.player.paused
        playerView.onPause()
    }

    override fun onChannelLoadStart(channelId: String) {
        if (BuildConfig.DEBUG) {
            Log.d("THEOliveView_$id", "onChannelLoadStart: $channelId")
        }
        isFirstPlaying = false
        mainScope.launch {
            flutterApi.onChannelLoadStart(channelId, emptyCallback)
        }
    }

    override fun onChannelLoaded(channelId: String) {
        if (BuildConfig.DEBUG) {
            Log.d("THEOliveView_$id", "onChannelLoaded: $channelId")
        }
        isFirstPlaying = false
        mainScope.launch {
            flutterApi.onChannelLoaded(channelId, emptyCallback)
        }
    }

    override fun onChannelOffline(channelId: String) {
        if (BuildConfig.DEBUG) {
            Log.d("THEOliveView_$id", "onChannelOffline: $channelId")
        }
        isFirstPlaying = false
        mainScope.launch {
            flutterApi.onChannelOffline(channelId, emptyCallback)
        }
    }

    override fun onWaiting() {
        if (BuildConfig.DEBUG) {
            Log.d("THEOliveView_$id", "onWaiting")
        }
        mainScope.launch {
            flutterApi.onWaiting(emptyCallback)
        }
    }

    override fun onPlay() {
        if (BuildConfig.DEBUG) {
            Log.d("THEOliveView_$id", "onPlay")
        }
        mainScope.launch {
            flutterApi.onPlay(emptyCallback)
        }
    }

    override fun onPlaying() {
        if (BuildConfig.DEBUG) {
            Log.d("THEOliveView_$id", "onPlaying")
        }
        if (!isFirstPlaying) {
            isFirstPlaying = true
        }

        mainScope.launch {
            flutterApi.onPlaying(emptyCallback)
        }
    }

    override fun onPause() {
        if (BuildConfig.DEBUG) {
            Log.d("THEOliveView_$id", "onPause")
        }
        mainScope.launch {
            flutterApi.onPause(emptyCallback)
        }
    }

    override fun onMutedChange() {
        if (BuildConfig.DEBUG) {
            Log.d("THEOliveView_$id", "onMutedChange")
        }
        mainScope.launch {
            flutterApi.onMutedChange(emptyCallback)
        }
    }

    override fun onIntentToFallback() {
        if (BuildConfig.DEBUG) {
            Log.d("THEOliveView_$id", "onIntentToFallback")
        }
        isFirstPlaying = false
        mainScope.launch {
            flutterApi.onIntentToFallback(emptyCallback)
        }
    }

    override fun onEnterBadNetworkMode() {
        if (BuildConfig.DEBUG) {
            Log.d("THEOliveView_$id", "onEnterBadNetworkMode")
        }
        mainScope.launch {
            flutterApi.onEnterBadNetworkMode(emptyCallback)
        }
    }

    override fun onExitBadNetworkMode() {
        if (BuildConfig.DEBUG) {
            Log.d("THEOliveView_$id", "onExitBadNetworkMode")
        }
        mainScope.launch {
            flutterApi.onExitBadNetworkMode(emptyCallback)
        }
    }

    override fun onReset() {
        if (BuildConfig.DEBUG) {
            Log.d("THEOliveView_$id", "onReset")
        }
        isFirstPlaying = false
        mainScope.launch {
            flutterApi.onReset(emptyCallback)
        }
    }

    override fun onError(message: String) {
        if (BuildConfig.DEBUG) {
            Log.d("THEOliveView_$id", "onError: $message")
        }
        isFirstPlaying = false
        mainScope.launch {
            flutterApi.onError(message, emptyCallback)
        }
    }

}

File: flutter_theolive_sdk_ios/ios/Classes/THEOliveView.swift
//
//  THEOliveView.swift
//  flutter_theolive
//
//  Created by Daniel on 16/07/2023.
//

import Foundation
import Flutter
import UIKit
import THEOliveSDK
import os

let log = OSLog(subsystem: Bundle.main.bundleIdentifier ?? "live.theo.THEOlive.Flutter" , category: "THEOliveView")

//TODO: This extension of Error is required to do use FlutterError in any Swift code.
//TODO: https://github.com/flutter/packages/blob/main/packages/pigeon/example/README.md#swift
extension FlutterError: Error {}

class THEOliveView: NSObject, FlutterPlatformView, THEOlivePlayerEventListener, THEOliveNativeAPI {

    private static var TAG = "FL_IOS_THEOliveView"
    private let _viewId: Int64
    private var _view: UIView
    private let _pigeonMessenger: PigeonMultiInstanceBinaryMessengerWrapper
    private let _flutterAPI: THEOliveFlutterAPI
    private let player = THEOliveSDK.THEOlivePlayer()
    private let emptyCompletion: (Result<Void, FlutterError>) -> Void = {result in }
    private var chromelessPlayerView: THEOliveSDK.THEOliveChromelessPlayerView?
    
    init(
        frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?,
        binaryMessenger messenger: FlutterBinaryMessenger?
    ) {
        _viewId = viewId
        _view = UIView()
        _view.frame = frame

        //setup pigeon
        _pigeonMessenger = PigeonMultiInstanceBinaryMessengerWrapper(with: messenger!, channelSuffix: "id_\(viewId)")
        _flutterAPI = THEOliveFlutterAPI(binaryMessenger: _pigeonMessenger)

        super.init()

        THEOliveNativeAPISetup.setUp(binaryMessenger: _pigeonMessenger, api: self)

        // iOS views can be created here
        createNativeView(view: _view)
        setupEventListeners()
    }

    func createNativeView(view _view: UIView){
        _view.backgroundColor = UIColor.black

        let chromelessPlayerView = THEOliveSDK.THEOliveChromelessPlayerView(player: player)
        chromelessPlayerView.translatesAutoresizingMaskIntoConstraints = false
        chromelessPlayerView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        chromelessPlayerView.frame = _view.bounds

        _view.addSubview(chromelessPlayerView)

        self.chromelessPlayerView = chromelessPlayerView
    }

    func setupEventListeners() {
        player.add(eventListener: self)
    }
    
    func view() -> UIView {
        return _view
    }

    func preloadChannels(channelIDs: [String]) throws {
        self.player.preloadChannels(channelIDs)
    }
    
    func loadChannel(channelID: String) throws {
        self.player.loadChannel(channelID)
    }

    func play() throws {
        self.player.play()
    }

    func pause() throws {
        self.player.pause()
    }

    func isAutoplay() throws -> Bool {
        return false // TODO: not implemented in native
    }
    
    func setMuted(muted: Bool) throws {
        self.player.muted = muted
    }
    
    func setBadNetworkMode(badNetworkMode: Bool) throws {
        // TODO: not implemented in native
    }
    
    func goLive() throws {
        self.player.goLive()
    }
    
    func reset() throws {
        self.player.reset()
    }

    func updateConfiguration(configuration: PigeonNativePlayerConfiguration) throws {
        let nativeConfig = THEOlivePlayer.Configuration(sessionId: configuration.sessionId)
        self.player.updateConfiguration(nativeConfig)
    }

    func dispose() throws {
        player.remove(eventListener: self)
        chromelessPlayerView?.removeFromSuperview()
        player.reset()
    }

    func onLifecycleResume() {
        // ignore on iOS
    }

    func onLifecyclePause() {
        // ignore on iOS
    }

    func onChannelLoadStart(channelId: String) {
        _flutterAPI.onChannelLoadStart(channelID: channelId, completion: emptyCompletion)
    }

    func onChannelLoaded(channelId: String) {
        _flutterAPI.onChannelLoaded(channelID: channelId, completion: emptyCompletion)
    }

    func onChannelOffline(channelId: String) {
        _flutterAPI.onChannelOffline(channelID: channelId, completion: emptyCompletion)
    }
    
    func onWaiting() {
        _flutterAPI.onWaiting(completion: emptyCompletion);
    }
    
    func onPlay() {
        _flutterAPI.onPlay(completion: emptyCompletion);
    }
    
    func onPlaying() {
        _flutterAPI.onPlaying(completion: emptyCompletion);
    }

    func onPause() {
        _flutterAPI.onPause(completion: emptyCompletion)
    }

    func onMutedChange() {
        _flutterAPI.onMutedChange(completion: emptyCompletion)
    }
    
    func onIntentToFallback() {
        _flutterAPI.onIntentToFallback(completion: emptyCompletion)
    }

//    func onEnterBadNetworkMode() {}
    
//    func onExitBadNetworkMode() {}
    
//    func onReset() {}
    
    func onError(message: String) {
        _flutterAPI.onError(message: message, completion: emptyCompletion)
    }

    deinit {
        os_log("deinit %d", log: log, type: .debug, _viewId)
    }
}
